edx-solutions/xblock-drag-and-drop-v2
# Move items between zones in assessment mode

**Description**

This PR implements the following use case:

> As a learner, I want to be able to move items between drop targets before checking (submitting) my answer in assessment mode. This should be possible for any item, regardless of whether I placed it on a correct or incorrect zone before. I also want to be able to drag items back to the item bank without having to use the reset button.

*UX*

The existing implementation does not allow learners to move an item to a different zone once it has been placed on the board. We will keep this behavior for drag-and-drop problems in standard mode.

In assessment mode, learners should be able to move items that have already been placed on the board, regardless of whether they are located in a correct or incorrect zone. To do this, learners will be able to use the same workflows that are already established for moving items from the item bank to the board:

- Items can be moved to a different zone using the mouse. Workflow:
  1. Pick item up by clicking it.
  2. Drag item to desired zone.
  3. Drop item.
- Items can be moved to a different zone using the keyboard. Workflow:
  1. (Shift-)TAB to item.
  2. Pick item up by pressing "Enter", "Space", "Ctrl-m", or "⌘-m".
  3. TAB to desired zone.
  4. Drop item by pressing "Enter", "Space", "Ctrl-m", or "⌘-m".
  5. Press "Esc" to cancel drop operation.

With no item selected, learners will be able to (Shift-)TAB to items that have been placed on the board.

Once an item has been picked up, pressing (Shift-)TAB will only navigate between zones and the item bank.

**JIRA tickets**: [SOL-1943](https://openedx.atlassian.net/browse/SOL-1943)

**Discussions**: [Discovery document for DnDv2 assessment mode](https://docs.google.com/document/d/1rFliZo1vlohJ7cde7KfP5bgEk0QqcvT1VNnUUa9_vjE/edit) (relevant section: 2.2.2)

**Dependencies**: None

**Sandbox URLs** (installed version de0312bfbba02b850c9032a6db63c728924505f5):

- LMS: http://dndv2-sandbox5.opencraft.hosting/
- Studio: http://studio-dndv2-sandbox5.opencraft.hosting/

DnD example is located at http://dndv2-sandbox5.opencraft.hosting/courses/course-v1:OpenCraftX+OC1+1/courseware/bdeaa08976fa4645a277edcaf5e8148c/e8ec2f5c98f4436abf532cccd6a71668/1?activate_block_id=block-v1%3AOpenCraftX%2BOC1%2B1%2Btype%40vertical%2Bblock%40e3c62b19090948b8ad89d3386bff4435

**Testing instructions**:

1. Add a DnDv2 block to a unit in Studio.
2. Click "EDIT" and change "Problem mode" to "assessment".
3. Drag items from the bank into drop zones. You may place some items in wrong zones.
4. Drag and drop already placed items between zones, both with mouse and keyboard.
5. Observe that the items "stick" in the placed position (are not moved back to the bank or previous zone). Observe that per-item feedback is not displayed.
6. Drag some items back to the bank with both mouse and keyboard.
6. Change "Problem mode" back to "standard".
7. Make sure everything still works correctly in standard mode.

**Author notes and concerns**:

If you put multiple items into the same zone with keyboard in assessment mode, you can only see the topmost item, the rest of the items are hidden behind it (the same thing can happen when dragging with the mouse, but when dragging with the mouse you usually don't drop the item in *exactly* the same place, so parts of overlapping items are usually still visible).

Using the existing zone alignment feature (where you can specify how dropped items automatically align inside zones) prevents this problem.

**Reviewers**
- [x] OpenCraft/Solutions: @haikuginger  
- [ ] a11y: @edx/edx-accessibility
